src/runtime_src/core/common/drv/xrt_cu.c

```diff
@@ -90,13 +117,45 @@ static inline void __process_sq(struct xrt_cu *xcu)
 		return;
 	}
 
-	/* Move all of the completed commands to completed queue */
-	while (xcu->done_cnt && xcu->num_sq) {
-		xcmd = list_first_entry(&xcu->sq, struct kds_command, list);
-		list_move_tail(&xcmd->list, &xcu->cq);
+	/* Sometimes a CU done but it doesn't ready for new command.
+	 * In this case, sq could be empty.
+	 */
+	if (!xcu->num_sq)
+		return;
+
+	BUG_ON(xcu->num_sq < xcu->done_cnt);
+
+	ev_client = first_event_client_or_null(xcu);
+	list_for_each_entry_safe(xcmd, next, &xcu->sq, list) {
+		if (xcu->done_cnt) {
+			/* Done commands have priority */
+			xcmd->status = KDS_COMPLETED;
+			--xcu->done_cnt;
+		} else if (unlikely(ev_client)) {
+			/* Client event happens rarely */
+			if (xcmd->client != ev_client)
+				continue;
+
+			tick = atomic_read(&xcu->tick);
+			/* Record CU tick to start timeout counting */
+			if (!xcmd->tick) {
+				xcmd->tick = tick;
+				xcmd->ttl = CU_SEC2TIMER(CU_EXEC_DEFAULT_TTL);
```

Comment:
do you really need xcmd->ttl? is it just a hard-coded value anyway and never change?

Reply:
I don't really need that for now. I would just use a hard-code value before command timeout is configurable.